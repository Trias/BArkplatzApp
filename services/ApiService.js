const ACCESS_TOKEN = '';
const URL = ``;

import dummyData from '../data/UpdatedDummy.json';

let cache = null;

let interval = null;

async function rpcCall(method, params = {}) {
    let response;
    try {
        console.log('start API call: '+ method + ' with params: ' + JSON.stringify(params));
        response = await fetch(URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                method, params
            })
        });
        let json = await response.json()

        console.log('api call successful');
        console.log('method: ' + method);
        console.log('params: ' + JSON.stringify(params));
        console.log(json);

        return json;
    } catch (e) {
        console.warn('error while getting API data')
        console.log(response);
        console.log(e);
    }
}

export async function initApi() {
    await fetchAllAsync();
    interval = setInterval(fetchAllAsync, 60 * 1000);
}
export function stopApi() {
    clearInterval(interval);
    interval = null;
}
export async function fetchAllAsync() {
    let response = await rpcCall('fetchAll');
    if (response && !response.error) {
        cache = addCoordinates(response)
    } else {
        console.warn('empty data from api, using old data');
    }
    return cache;
}
export async function getReservationsOfThisUserAsync() {
    return await rpcCall('getReservationsOfThisUser')
}
export async function reserveAsync(reservation) {
    return await rpcCall('reserve', reservation)
}
export async function deleteReservationAsync(reservation) {
    return await rpcCall('deleteReservation', reservation);
}
export async function getReservationOfOneParkingSpaceAsync(parkingSpaceId) {
    return await rpcCall('getReservationOfOneParkingSpace', { parkingSpaceId });
}
export function getData() {
    if (!cache) {
        console.warn('error fetching data from api, using dummy data');
        cache = addCoordinates(dummyData);
    }

    return cache;
}

function addCoordinates(cache){
    for (let parkingLot of Object.values(cache)) {
        let firstParkingSpace = Object.values(parkingLot.spaces)[0];
        parkingLot.latitude = firstParkingSpace.latitude;
        parkingLot.longitude = firstParkingSpace.longitude;
    }

    return cache
}
