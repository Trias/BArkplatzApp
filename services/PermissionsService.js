import * as Permissions from 'expo-permissions';

let hasLocationPermission = false;
export async function requestLocationPermission() {
    if (hasLocationPermission) {
        return true;
    }
    try {
        const { status } = await Permissions.askAsync(Permissions.LOCATION);
        if (status === 'granted') {
            hasLocationPermission = true;
            console.log('gps granted');
            return true;
        } else {
            console.log('gps permission denied');
            return false;
        }
    } catch (err) {
        console.warn(err);
        return false;
    }
}

let hasNotificationPermission = false;
export async function requestNotificationPermission() {
    if (hasNotificationPermission) {
        return true;
    }
    try {
        const { status } = await Permissions.askAsync(Permissions.USER_FACING_NOTIFICATIONS);
        if (status === 'granted') {
            hasNotificationPermission = true;
            console.log('notifications granted');
            return true;
        } else {
            console.log('notifications denied');
            return false;
        }
    } catch (err) {
        console.warn(err);
        return false;
    }
}