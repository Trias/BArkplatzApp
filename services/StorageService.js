import { AsyncStorage } from "react-native";

const cache = {};

export async function initStorage(){
    const allKeys = await AsyncStorage.getAllKeys();
    for (let key of allKeys){
        let value = await AsyncStorage.getItem(key);
        cache[key] = value;
    }
}

export function getItem(name){
    return cache[name];
}

export function setItem(name, item){
    cache[name] = item;
    setItemAsync(name, item);
}

export function removeItem(name){
    delete cache[name];
    removeItemAsync(name);
}

export async function getItemAsync(name){
    return AsyncStorage.getItem(name)
}

export async function setItemAsync(name, item){
    cache[name] = item;
    return AsyncStorage.setItem(name, item);
}

export async function removeItemAsync(name){
    cache[name] = undefined;
    return AsyncStorage.removeItem(name);
}