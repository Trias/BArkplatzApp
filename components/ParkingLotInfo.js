import React from 'react';
import {
    ScrollView,
    Text,
    Button,
    View,
    StyleSheet,
} from 'react-native';
import { UsageChart, Explanation } from './UsageChart';
import GetNavigation from './GetNavigation';
import { getData } from '../services/ApiService';

export function ParkingLotInfo({ parkingLotId , navigation}) {
    let parkingLot = getData()[parkingLotId];
    return <ScrollView style={styles.info}>
        <Text style={{ textAlign: "center", fontWeight: 'bold' }}>{parkingLot.name}</Text>
        <View style={{ flexDirection: "row", justifyContent: "space-around", paddingVertical: 5 }}>
            <Button
                title="Parkplatzansicht"
                onPress={() => {
                    navigation.navigate('ParkingLotDetail', { parkingLotId: parkingLot.id, title: parkingLot.name })
                }}
            />
            <GetNavigation coordinates={parkingLot} />
        </View>
        <Explanation parkingLot={parkingLot} />
        <UsageChart parkingLot={parkingLot} />
        <Text style={{ textAlign: "center" }}>Freie Parkpl√§tze: {parkingLot.unoccupied} / {Object.keys(parkingLot.spaces).length}</Text>
    </ScrollView>
}

const styles = StyleSheet.create({
    info: {
        flex: 1,
        height: '50%',
        paddingTop: 5,
        backgroundColor: '#fff',
    },
})