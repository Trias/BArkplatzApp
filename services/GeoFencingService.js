import * as TaskManager from 'expo-task-manager';
import * as Location from 'expo-location';
import { Notifications } from 'expo';
import { requestLocationPermission, requestNotificationPermission } from './PermissionsService';

const TASK_NAME = "parkingLots";

let geoFencingActive = false;
export async function initGeofencing() {
  try {
    // permissions returns only for location permissions on iOS and under certain conditions, see Permissions.LOCATION
    let locationStatus = await requestLocationPermission();
    let notificationStatus = await requestNotificationPermission();

    if (!geoFencingActive) {
      if (locationStatus && notificationStatus) {
        geoFencingActive = true;
        await Location.startGeofencingAsync(TASK_NAME, [{
          // erba: 49.903, 10.87
          // harcoded for now to avoid dependency on network request
          identifier: "631", // heinrichsdamm
          latitude: 49.89,
          longitude: 10.903,
          radius: 100
        },
        {
          identifier: "38", // brennerstraÃŸe
          latitude: 49.901,
          longitude: 10.90,
          radius: 100
        }]);
      } else {
        console.log('Location permission or notification permission not granted');
      }
    }
  } catch (e) {
    console.log('error while geofencing');
  }
}

TaskManager.defineTask(TASK_NAME, ({ data: { eventType, region }, error }) => {
  console.log("geofencing event");
  if (error) {
    // check `error.message` for more details.
    console.log(error)
    return;
  }
  if (eventType === Location.GeofencingEventType.Enter) {
    const localNotification = {
      title: "Sie sind am Parkplatz angekommen!",
      body: "jetzt Parkplatz merken"
    }
    Notifications.addListener(notification => {
      if (notification.origin == "selected") {
        let { MainTabNavigator } = import('../navigation/MainTabNavigator');
        MainTabNavigator.navigation.navigate('ParkingLotDetail', { parkingLotId: region.identifier })
      }
    })
    Notifications.presentLocalNotificationAsync(localNotification);
    console.log("You've entered region:", region);
  } else if (eventType === Location.GeofencingEventType.Exit) {
    console.log("You've left region:", region);
  }
});