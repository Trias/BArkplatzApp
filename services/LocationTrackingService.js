import * as Location from 'expo-location';
import {requestLocationPermission} from './PermissionsService';

let locationTrackingInterval = null;
let locationTrackingActive = false;

export async function initLocationTracking(){
  try {
    // permissions returns only for location permissions on iOS and under certain conditions, see Permissions.LOCATION
    let locationStatus = await requestLocationPermission();
    if (!locationTrackingActive && locationStatus){
      locationTrackingActive = true;
      locationCache = await getLocationAsync();
      locationTrackingInterval = setInterval(getLocationAsync, 10000);
    }
  }catch(e){
    console.log('error while location');
  }
}

export function stopLocationTracking(){
  clearInterval(locationTrackingInterval);
  locationTrackingInterval = null;
  locationTrackingActive = false;
}

let locationCache = null;
export async function getLocationAsync(){
  locationCache = await Location.getCurrentPositionAsync({enableHighAccuracy: true});
  return locationCache;
}

export function getLocation(){
  if(locationCache){
    return locationCache;
  }else{
    console.warn('could not get location data, using fallback');
    return {coords: {latitude: 49.9028939, longitude: 10.8686975}};
  }
}