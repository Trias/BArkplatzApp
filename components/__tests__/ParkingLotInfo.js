import React from 'react';
import renderer from 'react-test-renderer';

import {ParkingLotInfo} from '../ParkingLotInfo';

import * as ApiService from '../../services/ApiService';
import * as LocationTrackingService from '../../services/LocationTrackingService';

jest.mock('../../services/ApiService');
jest.mock('../../services/LocationTrackingService');

function addCoordinates(cache){
    for (let parkingLot of Object.values(cache)) {
        let firstParkingSpace = Object.values(parkingLot.spaces)[0];
        parkingLot.latitude = firstParkingSpace.latitude;
        parkingLot.longitude = firstParkingSpace.longitude;
    }

    return cache
}
ApiService.getData = jest.fn(() => {return addCoordinates(require('../../data/UpdatedDummy'))});

jest.useFakeTimers();

it(`renders correctly`, () => {
  const tree = renderer.create(<ParkingLotInfo 
    parkingLotId={'38'}
  />).toJSON();

  expect(tree).toMatchSnapshot();
});
