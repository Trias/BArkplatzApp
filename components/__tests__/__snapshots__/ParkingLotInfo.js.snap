// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": "50%",
      "paddingTop": 5,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Bamberg PuR Bahnhof Brennerstrasse 97
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingVertical": 5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Parkplatzansicht
          </Text>
        </View>
      </View>
      <View
        accessibilityRole="button"
        accessibilityStates={Array []}
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Navigation
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Es ist weniger los als sonst
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": 355,
          "marginLeft": 2,
          "paddingRight": 10,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 294,
            },
          ]
        }
      >
        <View
          onLayout={[Function]}
          style={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 10,
                "opacity": 0,
              }
            }
          >
            100%
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "marginLeft": 4,
            "marginRight": 8,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "height": 284,
              "width": 710,
            }
          }
        >
          <View
            onLayout={[Function]}
            style={
              Object {
                "flex": 1,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "marginTop": 2,
              "width": 710,
            }
          }
        >
          <View
            onLayout={[Function]}
            style={
              Object {
                "flexGrow": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": undefined,
                  "opacity": 0,
                }
              }
            >
              0-3
            </Text>
          </View>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Freie Parkpl√§tze: 
      85
       / 
      118
    </Text>
  </View>
</RCTScrollView>
`;
